### DISCUSSION TITLE
How to create filter on both tag and branch

### DISCUSSION THREAD

**Discussion by swenzel** [community/community#25244] [Actions]
<p>Hello everyone!</p>
<p>We’ve spent quite some time trying to figure out how to create a workflow that will only be run for commits that are both on master and tagged with <code>v*</code> in order to identify release commits.</p>
<p>The workflow we have in mind for creating releases is as follows:</p>
<ol>
<li>
<p>Checkout branch <code>release-prep</code></p>
</li>
<li>
<p>Merge the current <code>master</code> branch into it</p>
</li>
<li>
<p>Bump the version within the code</p>
</li>
<li>
<p>Update the <code>CHANGELOG.md</code> file</p>
</li>
<li>
<p>Commit the changes</p>
</li>
<li>
<p>Add a tag containing the version number (prefixed with v)</p>
</li>
<li>
<p>Push and open pull request to <code>master</code></p>
</li>
<li>
<p>Go through normal review workflow keeping the tag at the branch head</p>
</li>
</ol>
<p>Once everything is fine, the branch (along with the tag) will be merged into master which is then supposed to trigger publishing a release both to github and pypi.</p>
<p>If you’re interested, our project is:<br>
<a href="https://github.com/real-digital/esque/" rel="nofollow noopener">https://github.com/real-digital/esque/</a></p>
<p>The filter variants we’ve tried are:</p>
<pre><code>on:
push:
branches:
- master
tags:
- v*

on:
push:
tags:
- v*
branches-ignore:
- "*"
- "!master"

on:
  push:
    tags:
      - v*
    branches-ignore:
      - "!master"

on:
push:
branches:
- master
-"refs/tags/v\*"
</code></pre>
<p>Based on <a href="https://github.community/t/how-to-restrict-execution-of-github-actions-workflow-on-tags-creation-only/16041">https://github.community/t5/GitHub-Actions/How-to-restrict-execution-of-GitHub-Actions-workflow-on-tags/td-p/29567</a></p>
<p>We’re now falling back to a solution where the workflow itself fails when the conditions are not met. How do other people do this? I know there is a release draft option but there is so much you can miss in a release code wise that we’d like to enforce reviews and tests on it before it’s actually published.</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>Based on current paths/tags filter for on: push event, it doesn’t supported to filter a commit which are included in a release tag meanwhile in master branch.</p>
<p>There is a way to trigger a workflow when a new release published , using</p>
<p>on:</p>
<p>  release:</p>
<p>    types: [published]</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/9/9f1c65c60ebe0a00c72b8b68f255fc70a7a72eb7.png" data-download-href="/uploads/short-url/mHyL4Fq3SXabYNV2FopKM6sq8Xt.png?dl=1" title="publish .png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181089827-dd1cd28a-7843-4a54-954a-89cd2f2d5174.png" alt="publish .png" data-base62-sha1="mHyL4Fq3SXabYNV2FopKM6sq8Xt" width="495" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/9/9f1c65c60ebe0a00c72b8b68f255fc70a7a72eb7_2_495x500.png, https://aws1.discourse-cdn.com/github/optimized/2X/9/9f1c65c60ebe0a00c72b8b68f255fc70a7a72eb7_2_742x750.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/9/9f1c65c60ebe0a00c72b8b68f255fc70a7a72eb7.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/9/9f1c65c60ebe0a00c72b8b68f255fc70a7a72eb7_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">publish .png</span><span class="informations">772×779 27.9 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Please refer to this document: <a href="https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fhelp.github.com%2Fen%2Factions%2Fautomating-your-workflow-with-github-actions%2Fevents-that-trigger-workflows%23release-event-release&amp;data=02%7C01%7Cv-yjingz%40microsoft.com%7C53aa45b66733473e53e808d76da7c90a%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C637098439706591732&amp;sdata=VMcwuzWUYDqRW%2BkX%2F8QsUx%2F86fSLh6eymVv9Cvjpkww%3D&amp;reserved=0" rel="nofollow noopener">https://help.github.com/en/actions/automating-your-workflow-with-github-actions/events-that-trigger-workflows#release-event-release</a></p>

**Comment by swenzel** [comment]
<p>Hey, thanks for your answer!</p>
<p>That is actually what we do now. So once the release PR gets merged, we create a release manually. Once that release is published, it creates the tag and the corresponding pipeline pushes the package to pypi.</p>
<p>Guess there isn’t really a way to do it fully automated.</p>

**Comment by khatiwadanilav** [comment]
<p>Though I am replying to an old message, this is how I found the workaround.</p>
<p>Create a filter on both tag and branch is to compare the hash commit of the tag and master.</p>
<p>We can create a job that will check the hash commit. If the hash commit match than run another job to perform your desired actions. Else, if the hash commit don’t match you can make the workflow fail so that your next job doesn’t run.</p>
<p>This is how I do it.</p>
<pre><code>name: Check Deployment and versions

on:
  push:
    tags:
      - v1.*

jobs:
  check-tag-branch:
    runs-on: ubuntu-latest
    steps:
    - name: get tag commit hash
      id: tag-commit-hash
      run: |
        hash=${{ GITHUB.SHA }}
        echo "::set-output name=tag-hash::${hash}"
        echo $hash
    - name: checkout master
      uses: actions/checkout@v2
      with:
        ref: master
    - name: get latest master commit hash
      id: master-commit-hash
      run: |
        hash=$(git log -n1 --format=format:"%H")
        echo "::set-output name=master-hash::${hash}"
        echo $hash
    - name: exit if tag commit hash don't match master commit hash
      if: steps.tag-commit-hash.outputs.tag-hash != steps.master-commit-hash.outputs.master-hash
      run: exit 1

  build-base:
    needs: [check-tag-branch]
    runs-on: ubuntu-latest
    steps:
    - name: setup python
      uses: actions/setup-python@v1
    - name: echo something
      run: echo "This steps run successfully"
</code></pre>

**Comment by rickstaa** [comment]
<p>Based on the answer of <a class="mention" href="/khatiwadanilav">@khatiwadanilav</a> I created an action which currently using in my workflows to only trigger jobs when a tag was pushed to a specific branch:</p>
<pre><code class="lang-yaml">name: Github tag test
on:
  push:
    tags:
      - "v*.*.*"
jobs:
  on-main-branch-check:
    runs-on: ubuntu-latest
    outputs:
      on_main: ${{ steps.contains_tag.outputs.retval }}
    steps:
      - uses: actions/checkout@v2
        with:
          fetch-depth: 0
      - uses: rickstaa/action-contains-tag@v1
        id: contains_tag
        with:
          reference: "main"
          tag: "${{ github.ref }}"
  tag-on-main-job:
    runs-on: ubuntu-latest
    needs: on-main-branch-check
    if: ${{ needs.on-main-branch-check.outputs.on_main == 'true' }}
    steps:
      - run: echo "Tag was pushed to main."
      - run: echo ${{needs.on-main-branch-check.outputs.on_main}}
  tag-not-on-main-job:
    runs-on: ubuntu-latest
    needs: on-main-branch-check
    if: ${{ needs.on-main-branch-check.outputs.on_main != 'true' }}
    steps:
      - run: echo "Tag was not pushed to main."
      - run: echo ${{needs.on-main-branch-check.outputs.on_main}}
</code></pre>
<p>You can see this workflow in action <a href="https://github.com/rickstaa/action-test-repo/runs/1708005051?check_suite_focus=true" rel="noopener nofollow ugc">here</a>. I also send a <a href="https://support.github.com/contact/feedback" rel="noopener nofollow ugc">feature request to the GitHub support</a> to support this without using a workaround. Hope GitHub adds this feature soon. In the meantime, feel free to create <a href="https://github.com/rickstaa/action-contains-tag/" rel="noopener nofollow ugc">an issue or a pull request</a> onto my workaround if something is missing.</p>

**Comment by benipeled** [comment]
<p>Thanks <a class="mention" href="/rickstaa">@rickstaa</a> for sharing this workaround,<br>
Can you please share a link to the submitted feature request?</p>

**Comment by rickstaa** [comment]
<p><a class="mention" href="/benipeled">@benipeled</a> At the time of writing, I could not find a public place to submit the feature request. I therefore followed the advice of <a href="https://github.community/t/where-do-i-post-feature-requests-for-github/1501/22" class="inline-onebox">Where do I post feature requests for Github ? - #22 by AndreaGriffiths11</a> and opened a private support ticket. I did not receive a response on that ticket. Feel free to create a feature request somewhere else if there is now a public place where a feature request can be submitted.</p>

**Comment by benipeled** [comment]
<p><a class="mention" href="/rickstaa">@rickstaa</a>, yes they now call it <em>discussions</em> - submitted [0],</p>
<p>btw, github recently released (public-beta) the “Protected Tags” feature [1] which might be helpful for this case by controlling who can push tags (admin &amp; maintainers).</p>
<p>[0] <a href="https://github.com/github/feedback/discussions/13226" class="inline-onebox" rel="noopener nofollow ugc">Trigger workflow on push-tag event to a specific branch · Discussion #13226 · github/feedback · GitHub</a><br>
[1] <a href="https://github.com/github/feedback/discussions/10906" class="inline-onebox" rel="noopener nofollow ugc">🆕 Private beta: Protected tags · Discussion #10906 · github/feedback · GitHub</a></p>

**Comment by rickstaa** [comment]
<p><a class="mention" href="/benipeled">@benipeled</a> Ah, thanks a lot for letting me know about the discussions and the “Protected Tags” feature!</p>
